### ISSUE TITLE
Villagers don't trade

### ISSUE THREAD

**Issue by yangm97** [cuberite/cuberite#3450]
Taken from #1109

**Comment by Seadragon91** [comment]
I'm trying my luck :)

**Comment by madmaxoft** [comment]
Consider consulting early, especially the way you plan on storing the available trades, plugin interaction etc.

**Comment by Seadragon91** [comment]
My current idea is:
* The `cTradeWindow` for the 3 slots (input 1 (0), input 2 (1) and output (2)
* A class named `cSingleTrade` that contains all the info for one trade
* A class named `cMerchant` that contains a list of `cSingleTrade`

The class `cSingleTrade` would contain this information:
* Input item 1
* Output item
* Input item 2 is optional
* Is Trade disabled
* Max amount it can be done
* Current amount it has been done

Used this link to create the list above. The info are send to the player, when the trade window is opened.
http://wiki.vg/Plugin_channels#MC.7CTrList

The class `cMerchant` would contain methods to:
* Add, update trades
* Passing a player to open a trade - vanilla only allows one trade per villager. Should we allow multiple players? I see there possible problems: race condition and desync - amount of trades / Is it disabled?
* This class would be exported to lua. The plugin dev would need to keep a instance of it in a list / global, to avoid garbage collection

Will for now checking this one:
http://wiki.vg/Plugin_channels#MC.7CTrSel

**Comment by Pokechu22** [comment]
One other thing (that I don't know the details of): in vanilla, you can use _any_ trade in the villager from the first page; if you put the required items in, it'll give you the result items, always.  But IIRC if you change the page that no longer is the case.  This is actually a really nice behavior, because it reduces clicking.

**Comment by Seadragon91** [comment]
I see a difficulty... . What if is the input and the ouput items of trades are the same? How can I distinguish them?  :)

**Comment by Pokechu22** [comment]
I think in that case, it uses the first applicable trade.  But I don't know the details.  There's also something about swapping the first and second items, but I can't remember if it's "that is allowed" or "that is not allowed" or "that is allowed sometimes".

**Comment by madmaxoft** [comment]
I'd rename `cSingleTrade` to `cTradeRecipe`. Otherwise this seems good.
Somehow the `cMerchant` class will need to keep track of past trades (per player?) to remove those used too many times, and unlock new ones. Or perhaps make it a yet separate `cVillagerTradeManager` class that will have hook-like access to trades made, and will modify the `cMerchant`'s recipes.

I don't understand about "one trade per villager" - does that mean that while player A has a villager's trade window open, player B cannot open that villager's trade window? Makes sense, somewhat, and we could do it like this for starters, and only improve on it later on.

The `cMerchant` class should not be GCed by Lua; rather, it should be owned by the individual villager, accessed via their Lua API, and never directly constructed by Lua. Even the `cTradeRecipe` class needn't be exported at all, as long as the `cMerchant` bindings are done properly to take everything in separate parameters (`AddTrade(cItem a_Input1, cItem a_Input2, cItem a_Output, ...)`) and return things in multiple return values (`GetAllTrades()` -> `{ {Input1, Input2, Output, ...}, {Input1, Input2, Output, ...}, ...}` etc.)

**Comment by Pokechu22** [comment]
> I don't understand about "one trade per villager" - does that mean that while player A has a villager's trade window open, player B cannot open that villager's trade window? Makes sense, somewhat, and we could do it like this for starters, and only improve on it later on.

Yes.

**Comment by Seadragon91** [comment]
I think it should be possible to use the class `cMerchant` without the need of a villager, e.g: Clicking a sign

**Comment by madmaxoft** [comment]
Yes, exactly for that reason it should be a separate class, rather than built into a `cVillager`. And for signs, it would be nice to be able to set fixed trades and not worry about trades being overused or adding new trades, like a villager does. That's why the `cVillagerTradeManager` should be separated from `cMerchant`.

**Comment by Seadragon91** [comment]
Okay read the infor that I get wrong. It's the current trade slot number - index of the array. With this now I know what the player is trading and the problem with same input / output item is gone.
